arctic/async/async_arctic.py

```diff
@@ -0,0 +1,187 @@
+                                                                            AsyncRequestType.MODIFIER))
+
+            # Create the request object
+            request = AsyncRequest(kind, lib_name, fun, *args, **kwargs)
+
+            # Update the state of tracked tasks
+            self._add_request(request)
+
+            # Submit the task
+            try:
+                new_id, new_future = self.submit_task(False, _arctic_task_exec, request)
+                request.id = new_id
+                request.future = new_future
+            except Exception:
+                # clean up the state
+                self._remove_request(request)
+                raise
+
+        # No need to hold the lock for the below statements
+        # If request is already finished by now, the callback is invoked immediately
+        request.future.add_done_callback(lambda the_future: self._request_finished(request, callback))
+
+        return request
+
+    def _request_finished(self, request, callback=None):
+        with type(self)._POOL_LOCK:
+            self._remove_request(request)
+        request.is_completed = True
+        if callback:
+            callback(request)
+
+    @staticmethod
+    def wait_request(request, timeout=None):
+        while request is not None and not request.is_completed:
+            AsyncArctic.wait_tasks_or_abort((request.future, ), timeout=timeout)
+
+    @staticmethod
+    def wait_requests(requests, timeout=None):
+        while requests and not all(r.is_completed for r in requests):
+            AsyncArctic.wait_tasks_or_abort(tuple(r.future for r in requests if not r.is_completed), timeout=timeout)
+
+    @staticmethod
+    def wait_any_request(requests, timeout=None):
+        while requests and not any(r.is_completed for r in requests):
+            AsyncArctic.wait_tasks(tuple(r.future for r in requests if not r.is_completed),
+                                   timeout=timeout, return_when=FIRST_COMPLETED, raise_exceptions=True)
+
+    @staticmethod
+    def filter_finished_requests(requests, do_raise=True):
+        if not requests:
+            return requests, requests
+        alive_requests = [r for r in requests if not r.is_completed]
+        done_requests = [r for r in requests if r.is_completed]
+        if do_raise:
+            AsyncArctic.raise_errored(done_requests)
+        return alive_requests, done_requests
+
+    @staticmethod
+    def raise_errored(requests):
+        errored = tuple(r for r in requests if r.is_completed and r.exception is not None)
+        if errored:
+            raise errored[0].exception
+
+    def join(self, timeout=None):
+        while len(self.requests_by_id):
+            try:
+                AsyncArctic.wait_requests(self.requests_by_id.values(), timeout=timeout)
+            except Exception as e:
+                logging.exception("Failed to join all requests")
+
+    def total_requests(self):
```

Comment:
total_requests isn't the right name, pending_requests?

Reply:
good one